### PR TITLE
Controller Wars 6: Return of the Force Controller

### PR DESCRIPTION
**Pull request by brndd**
**BEGIN DISCLAIMER**

noidea

**END DISCLAIMER**

I was told the main issue with force controllers was player movement, so I took a force controller branch that @metalgearsloth had worked on and fixed the player movement in it by making it use a fluid drag model very common in video games. The movement feels good now but I don't know anything about the rest of whatever the fuck force controllers might be needed for and I don't know if my design is gonna cause trouble I haven't foreseen because of reasons in the disclaimer above.

I'm putting this PR here so maybe others can look at it. It involves engine changes (all of which are courtesy of @metalgearsloth and I don't even know what they are or where or what they do) and I'm not putting those into the engine repo until this matures, so if you want to try it pull this branch: https://github.com/brndd/RobustToolbox/tree/movement_shit

# Examples

## High drag

This is what you would realistically normally use under normal realistic conditions. Players accelerate and decelerate quickly, in this example 0.2 seconds to reach max speed which is so fast you can't even tell there's any delay. https://media.discordapp.net/attachments/310555209753690112/771934734037024798/highdrag.mp4

## Low drag

This is what you use when you want floaty movement. Could be used for a low friction effect (ice level) etc. though right now the drag coefficient is hard-coded. https://media.discordapp.net/attachments/310555209753690112/771934737081696266/lowdrag.mp4

# Technical details

Basically see this StackExchange question: https://stackoverflow.com/questions/667034/simple-physics-based-movement

In physics terms we're using what's called "Stokes' drag". Normally it applies to fluid drag like air resistance and such, but video game physics aren't real life and using it for player movement is very common. It's *only* used in the player movement controller to enforce an upper bound on player speed and to create gradual acceleration, and not in the underlying physics system, so there shouldn't be any effect on stuff like throwing players around.

I also added in a kind of a stupid hack to counteract for the standard friction between a human and the average station floor so that the movement speed is closer to what we want it to be. As a result of this, if the actual friction is lower then you'll also move faster and if it's higher you'll move slower. It's far from realistic right now because you'll have the same traction in a zero-friction situation as you would in a very high friction situation... if that turns out to be an issue I'll have to figure something out, but it's not like the pre-existing movement uses any sort of realistic friction either.

### CODE REVIEW

File: Content.Shared/GameObjects/EntitySystems/SharedMoverSystem.cs [hunk 1]
```diff
@@ -4,6 +4,7 @@
 using Content.Shared.GameObjects.Components.Movement;
 using Content.Shared.Physics;
 using Content.Shared.Physics.Pull;
+using NFluidsynth;
```
Comment (Zumorica) [inline Content.Shared/GameObjects/EntitySystems/SharedMoverSystem.cs:7]: Happens to the best of us.

**Comment by brndd** [comment]
I have now merged in the old force controller work by ... everyone who worked on it. Seems like a good starting point though maybe it isn't.

I don't like working on the controllers that aren't player movement so really hoping someone else can do all the hard work for me.

File: Content.Server/Atmos/HighPressureMovementController.cs [hunk 2]
```diff
@@ -48,38 +46,25 @@ public class HighPressureMovementController : VirtualController
...
-
                 if (maxForce > ThrowForce)
                 {
                     if (throwTarget != EntityCoordinates.Invalid)
                     {
-                        var moveForce = maxForce * MathHelper.Clamp(moveProb, 0, 100) / 150f;
+                        var moveForce = maxForce * MathHelper.Clamp(moveProb, 0, 100) / 25;
                         var pos = ((throwTarget.Position - transform.Coordinates.Position).Normalized + direction.ToDirection().ToVec()).Normalized;
-                        LinearVelocity = pos * moveForce;
+                        ControlledComponent.Force += pos * moveForce * ControlledComponent.Mass * 1/frameTime;
```
Comment (brndd) [inline Content.Server/Atmos/HighPressureMovementController.cs:55]: > * 1/frameTime

:thinking:

**Comment by brndd** [comment]
*ahem* fuck prediction and fuck our physics

thanks for coming to my TED talk

**Comment by metalgearsloth** [comment]
Another one bites the dust.

**Comment by metalgearsloth** [comment]
Anyone still involved in this want to look at my branch https://github.com/metalgearsloth/space-station-14/tree/force_mover I think I'm pretty close to what we need. I've only looked at addressing player movement so everything else is essentially turned off.

* Fixes weightless reconciliation issue
* Warmstarts (doesn't reset velocity every tick for player)
* Low tickrates slightly worse than currently (bleh), I'm talking sub-20.
* The only other issue left is I need an artificial drag when the player stops moving but I realise the issue is that the client is calling StopMoving every frame whereas the server only calls it every tick which I need to use subtick for so they don't slide so much when stopping input. I don't want to increase friction as it'll make turning feel worse (though values subject to change).

**Comment by brndd** [comment]
Alright, this is hella fucking frustrating.

The new movement I had *works*. If I switch it over to using regular LinearVelocity instead of the new "BiasedLinearVelocity" that needs to be added for this, it works perfectly. Prediction is fine, subticking is fine, it all just works.

The problem is that in order for physics as a whole to generally work sanely, LinearVelocity has to be split into two components. I'm calling these components LinearVelocity, which "warm starts" ie. doesn't reset every tick, and BiasedLinearVelocity, which resets every tick and is where snowflake physics stuff that needs to directly modify velocity goes (read: movement and conveyors). I'm calling it this because that's what it's called in Godot, which seems to be doing the same thing.

The problem is, *something* about the way I implemented this new BiasedLinearVelocity is completely fucking breaking prediction. Movement gets jank if I switch the controller over to it. Movement is fine if it uses the old LinearVelocity, even though everything else about the movement controller remains the same. I don't understand why. 

I've tried going through every single usage of LinearVelocity and making it aware of BiasedLinearVelocity but to no avail. I don't understand why it seems to be getting different treatment from the pre-existing LinearVelocity for no good reason.

**Comment by brndd** [comment]
Sloth has better ideas for physics reworks than this so let's can it.

**Pull request closed without merging**